tket/src/Gate/Gate.cpp

```diff
@@ -221,13 +225,79 @@ Op_ptr Gate::transpose() const {
   }
 }
 
+static bool params_contain_nan(const std::vector<Expr>& params) {
+  static const std::regex nan_regex("\\bnan\\b");
+  for (const Expr& e : params) {
+    stringstream ss;
+    ss << e;
+    if (std::regex_search(ss.str(), nan_regex)) return true;
+  }
+  return false;
+}
+
+static double random_perturbation() {
+  static RNG rng;
+  int a = rng.get_size_t(20);
+  return (a - 10) * EPS;
```

Comment:
We might want to consider dropping the range down a bit. In the ideal world, our compilation has deterministic results, but even in this case we should aim for angles to not differ by more than `ERR_EPS` between runs. Whilst `ERR_EPS` is 10*`EPS`, these permutations could be added to multiple symbols in the expression and I don't know whether the trig will blow this up to a larger error. Probably a bit much to do the calculations by hand to see what a viable range would be, but if we can get away with halving the range that might be good

Reply:
Done.